### PR TITLE
Exhaustiveness checks for Opus::Enum

### PR DESCRIPTION
**Pull request by jez**
### Motivation

This PR allows for:

- exhaustiveness over the cases within an Opus::Enum
- Opus::Enum values to be used in type positions

### Test plan

See included automated tests.

### CODE REVIEW

File: core/Names.cc [hunk 1]
```diff
@@ -121,6 +124,10 @@ string Name::show(const GlobalState &gs) const {
                 return absl::StrCat(this->unique.original.data(gs)->show(gs), " (overload.", this->unique.num, ")");
             } else if (this->unique.uniqueNameKind == UniqueNameKind::MangleRename) {
                 return fmt::format("{}${}", this->unique.original.data(gs)->show(gs), this->unique.num);
+            } else if (this->unique.uniqueNameKind == UniqueNameKind::OpusEnum) {
+                // Mentioning this explicitly for grepping: the entire goal of UniqueNameKind::OpusEnum is
+                // to have Name::show print the name as if on the original name, so that our OpusEnum
+                // DSL-synthesized class names are kept as an implementation detail.
```
Comment (elliottt) [inline core/Names.cc]: Could you modify this comment to state explicitly that you're falling through here?

File: dsl/OpusEnum.cc [hunk 2]
```diff
@@ -38,34 +40,104 @@ bool isOpusEnum(core::MutableContext ctx, ast::ClassDef *klass) {
...
+            return {};
         }
 
-        auto loc = asgn->rhs->loc;
-        asgn->rhs = ast::MK::Send2(loc, ast::MK::T(loc), core::Names::let(), move(asgn->rhs), ast::MK::Self(loc));
+        if (!(arg0->fun == core::Names::new_() && arg0->recv->isSelfReference())) {
+            return {};
+        }
     }
+
```
Comment (elliottt) [inline dsl/OpusEnum.cc]: There's a ton of nested pattern-matching happening above, could you put a short comment that describes what you expect the current expression to look like at this point?

File: dsl/OpusEnum.cc [hunk 3]
```diff
@@ -38,34 +40,104 @@ bool isOpusEnum(core::MutableContext ctx, ast::ClassDef *klass) {
...
+    auto name = ctx.state.freshNameUnique(core::UniqueNameKind::OpusEnum, lhs->cnst, 1);
+    auto classCnst = ast::MK::UnresolvedConstant(lhs->loc, ast::MK::EmptyTree(), name);
+    ast::ClassDef::ANCESTORS_store parent;
+    parent.emplace_back(klass->name->deepCopy());
+    ast::ClassDef::RHS_store classRhs;
+    classRhs.emplace_back(ast::MK::Send1(stat->loc, ast::MK::Self(stat->loc), core::Names::extend(),
+                                         ast::MK::Constant(stat->loc, core::Symbols::T_Helpers())));
+    classRhs.emplace_back(ast::MK::Send1(stat->loc, ast::MK::Self(stat->loc), core::Names::include(),
+                                         ast::MK::Constant(stat->loc, core::Symbols::Singleton())));
+    classRhs.emplace_back(ast::MK::Send0(stat->loc, ast::MK::Self(stat->loc), core::Names::declareFinal()));
```
Comment (elliottt) [inline dsl/OpusEnum.cc]: Will it be a problem that it introduces these three for every assignment? Would it be better to set a flag, and add them in `postTransformClassDef`?

Comment (elliottt) [reply dsl/OpusEnum.cc]: Never mind, this isn't going onto the class being examined, but the one being generated.

File: resolver/type_syntax.cc [hunk 4]
```diff
@@ -559,6 +559,23 @@ TypeSyntax::ResultType TypeSyntax::getResultTypeAndBind(core::MutableContext ctx
...
+            // non-alias constant for an instance of a particular Opus::Enum--it was created with
+            // `new` and immediately assigned into a constant. Any further ConstantLits of this enum's
+            // type must be either class aliases (banned in type syntax) or type aliases.
+            //
+            // This is not the case for arbitrary singletons: MySingleton.instance can be called as many
+            // times as wanted, and assigned into different constants each time. As much as possible, we
+            // want there to be one name for every type; making an alias for a type should always be
+            // syntactically declared with T.type_alias.
+            if (auto resultType = core::cast_type<core::ClassType>(maybeAliased.data(ctx)->resultType.get())) {
+                if (resultType != nullptr &&
```
Comment (elliottt) [inline resolver/type_syntax.cc]: Isn't `resultType != nullptr` true already, as you're underneath the `if` where it's defined with `cast_type`?

File: core/types/calls.cc [hunk 5]
```diff
@@ -307,6 +307,12 @@ TypePtr unwrapType(Context ctx, Loc loc, const TypePtr &tp) {
         return metaType->wrapped;
     }
     if (auto *classType = cast_type<ClassType>(tp.get())) {
+        if (classType->symbol.data(ctx)->derivesFrom(ctx, core::Symbols::OpusEnum())) {
+            // Opus::Enum instances are allowed to stand for themselves in type syntax positions.
+            // See the note in type_syntax.cc regarding Opus::Enum.
+            return tp;
+        }
+
```
Comment (jez) [inline core/types/calls.cc]: @DarkDimius I made `unwrapType` in calls.cc aware of Opus::Enum. I'm not sure if this is going too far.

**Comment by jez** [comment]
Marking WIP while I fix the pay-server errors.

File: core/types/calls.cc (hunk 5 above)
Comment (DarkDimius) [reply core/types/calls.cc]: could you add the following test:
```
class Enum < Opus::Enum
  Entry = new
end

Enum::Entry.new # we should error here
T.cast(nil, T::Array[Enum::Entry]) # should be allowed
```

**Pull request merged**